.gitlab-ci.yml

```diff
@@ -29,22 +29,56 @@ variables:
     paths:
       - artifacts/
 
-test-linux:
+test-audit:
   stage:                           test
-  variables:
-    RUN_TESTS:                     all
   script:
-    - scripts/gitlab/test-all.sh
+    - set -e
+    - set -u
+    - cargo audit
+  tags:
+    - linux-docker
+
+test-validate-chainspecs:
+  stage:                           test
+  script:
+    - ./scripts/gitlab/validate_chainspecs.sh
     - sccache -s
   tags:
     - linux-docker
 
-test-audit:
+cargo-check:
   stage:                           test
   script:
-    - set -e
-    - set -u
-    - cargo audit
+    - ./scripts/gitlab/cargo_check.sh
+    - sccache -s
+  tags:
+    - linux-docker
+
+test-rust-changes:
+  stage:                           test
+  script:
+    - ./scripts/gitlab/rust_changes.sh
+  tags:
+    - linux-docker
+  allow_failure:                   true
+
+test-cpp:
+  stage:                           build
+  dependencies:
+    - test-rust-changes
+  script:
+    - ./scripts/gitlab/test-cpp.sh
+    - sccache -s
```

Comment:
what does this do and why is it run only after the building step?

Reply:
that's for debug purposes here, when we'll introduce `sccache` via docker volumes it will help.
The command itself shows what's going on with caches.

Reply:
and in here, `test-cpp.sh` does not use cargo, I understand your point.
In my opinion it should be complied with cargo though.